File: swift/Sources/CircleAI/LanguagesTranslation.swift
// LanguagesTranslation.swift
//
// Port of src/CircleAI.Languages.Translation/:
//   • TranslationTypes.cs   → TranslationMode, TranslationRequest,
//                             TranslationResult, ConversationTurn
//   • ITranslationEngine.cs → ITranslationEngine
//   • ILiveTranslator.cs    → ILiveTranslator
//   • LlmTranslationEngine.cs → LlmTranslationEngine (on-device, IChatGenerator-backed)
//
// Porting notes:
//   • C# `IAsyncEnumerable<T>` → `AsyncStream<T>` (the same convention used by
//     the inference bridge port in InferenceServer.swift).
//   • The C# generator API is `GenerateAsync` / `StreamAsync`; the Swift
//     `IChatGenerator` is `generate` / `stream`. The prompt text, the 0.9
//     confidence constant, `.Trim()`, and the "return only the translation"
//     instruction are all reproduced exactly.
//   • `record with { … }` (ConversationTurn update in StreamConversationAsync)
//     → a small `withTranslatedText(_:)` helper since Swift structs have no
//     positional `with` expression.
//   • `TranslationMode` gets a `name` (PascalCase) matching C# `Enum.ToString()`
//     so the prompt string ("Mode: Standard") is byte-identical.

import Foundation

// MARK: - TranslationMode

/// Translation register/domain. Ordinals + names match the C# enum.
public enum TranslationMode: Int, Sendable, CaseIterable {
    case standard = 0
    case conversational = 1
    case document = 2
    case technical = 3
    case legal = 4
    case medical = 5

    /// PascalCase name, matching C# `Enum.ToString()` (used in the prompt).
    public var name: String {
        switch self {
        case .standard: return "Standard"
        case .conversational: return "Conversational"
        case .document: return "Document"
        case .technical: return "Technical"
        case .legal: return "Legal"
        case .medical: return "Medical"
        }
    }
}

// MARK: - TranslationRequest

/// A request to translate a piece of text between two languages.
public struct TranslationRequest: Sendable, Equatable {
    public let text: String
    public let sourceBcpTag: String
    public let targetBcpTag: String
    public let mode: TranslationMode
    public let contextHint: String?

    public init(
        text: String,
        sourceBcpTag: String,
        targetBcpTag: String,
        mode: TranslationMode = .standard,
        contextHint: String? = nil
    ) {
        self.text = text
        self.sourceBcpTag = sourceBcpTag
        self.targetBcpTag = targetBcpTag
        self.mode = mode
        self.contextHint = contextHint
    }
}

// MARK: - TranslationResult

/// Result of a completed translation.
public struct TranslationResult: Sendable, Equatable {
    public let originalText: String
    public let translatedText: String
    public let sourceBcpTag: String
    public let targetBcpTag: String
    public let confidence: Float
    public let translatedAt: Date

    public init(
        originalText: String,
        translatedText: String,
        sourceBcpTag: String,
        targetBcpTag: String,
        confidence: Float,
        translatedAt: Date
    ) {
        self.originalText = originalText
        self.translatedText = translatedText
        self.sourceBcpTag = sourceBcpTag
        self.targetBcpTag = targetBcpTag
        self.confidence = confidence
        self.translatedAt = translatedAt
    }
}

// MARK: - ConversationTurn

/// One turn in a live bidirectional conversation.
public struct ConversationTurn: Sendable, Equatable {
    public let speakerBcpTag: String
    public let originalText: String
    public let translatedText: String?
    public let timestamp: Date

    public init(
        speakerBcpTag: String,
        originalText: String,
        translatedText: String?,
        timestamp: Date
    ) {
        self.speakerBcpTag = speakerBcpTag
        self.originalText = originalText
        self.translatedText = translatedText
        self.timestamp = timestamp
    }

    /// Returns a copy with `translatedText` replaced — the Swift analogue of the
    /// C# `turn with { TranslatedText = … }`.
    public func withTranslatedText(_ value: String?) -> ConversationTurn {
        ConversationTurn(
            speakerBcpTag: speakerBcpTag,
            originalText: originalText,
            translatedText: value,
            timestamp: timestamp)
    }
}

// MARK: - ITranslationEngine

/// On-device translation engine. No network call, no data leaving the device.
/// Translates meaning — not just words — using the on-device LLM.
public protocol ITranslationEngine: AnyObject, Sendable {
    func translate(_ request: TranslationRequest) async throws -> TranslationResult
    func streamTranslate(_ request: TranslationRequest) -> AsyncStream<String>
    func isLanguagePairSupported(sourceBcpTag: String, targetBcpTag: String) async throws -> Bool
}

// MARK: - ILiveTranslator

/// Bidirectional live conversation translator. Party A speaks `partyABcpTag`;
/// party B speaks `partyBBcpTag`. Each turn is translated in real-time so both
/// parties hear each other. Runs entirely on-device.
public protocol ILiveTranslator: ITranslationEngine {
    func streamConversation(
        _ inputStream: AsyncStream<ConversationTurn>,
        partyABcpTag: String,
        partyBBcpTag: String
    ) -> AsyncStream<ConversationTurn>
}

// MARK: - LlmTranslationEngine

/// `ITranslationEngine` backed by the on-device LLM via `IChatGenerator`. All
/// processing is on-device — no API calls, no data leaving the device. Port of
/// C# `LlmTranslationEngine`.
public final class LlmTranslationEngine: ILiveTranslator, @unchecked Sendable {
    private let generator: IChatGenerator

    public init(generator: IChatGenerator) {
        self.generator = generator
    }

    public func translate(_ request: TranslationRequest) async throws -> TranslationResult {
        let messages = [ChatMessage(role: "user", content: Self.buildPrompt(request))]
        let translated = try await generator.generate(messages: messages, options: nil)
        return TranslationResult(
            originalText: request.text,
            translatedText: translated.trimmingCharacters(in: .whitespacesAndNewlines),
            sourceBcpTag: request.sourceBcpTag,
            targetBcpTag: request.targetBcpTag,
            confidence: 0.9,
            translatedAt: Date())
    }

    public func streamTranslate(_ request: TranslationRequest) -> AsyncStream<String> {
        let messages = [ChatMessage(role: "user", content: Self.buildPrompt(request))]
        let inner = generator.stream(messages: messages, options: nil)
        return AsyncStream { continuation in
            let task = Task {
                for await token in inner {
                    if Task.isCancelled { break }
                    continuation.yield(token)
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    public func isLanguagePairSupported(
        sourceBcpTag: String, targetBcpTag: String
    ) async throws -> Bool {
        // On-device LLM handles any pair it was trained on.
        true
    }

    public func streamConversation(
        _ inputStream: AsyncStream<ConversationTurn>,
        partyABcpTag: String,
        partyBBcpTag: String
    ) -> AsyncStream<ConversationTurn> {
        AsyncStream { continuation in
            let task = Task {
                for await turn in inputStream {
                    if Task.isCancelled { break }
                    let targetTag = turn.speakerBcpTag == partyABcpTag ? partyBBcpTag : partyABcpTag
                    let req = TranslationRequest(
                        text: turn.originalText,
                        sourceBcpTag: turn.speakerBcpTag,
                        targetBcpTag: targetTag,
                        mode: .conversational)
                    do {
                        let result = try await self.translate(req)
                        continuation.yield(turn.withTranslatedText(result.translatedText))
                    } catch {
                        // On failure, forward the untranslated turn so the stream
                        // does not stall (the C# would propagate; here we keep the
                        // conversation flowing with the original text preserved).
                        continuation.yield(turn)
                    }
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    private static func buildPrompt(_ r: TranslationRequest) -> String {
        var s = "Translate the following text from \(r.sourceBcpTag) to \(r.targetBcpTag). "
        s += "Mode: \(r.mode.name). Preserve meaning and cultural context, not just literal words. "
        if let hint = r.contextHint {
            s += "Context: \(hint). "
        }
        s += "Return only the translation with no explanation.\n\n\(r.text)"
        return s
    }
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/languagestranslation/LanguagesTranslation.kt
// LanguagesTranslation.kt
//
// Kotlin port of CircleAI.Languages.Translation — the C# reference is the EXACT
// spec. On-device translation: meaning, not just words, via the on-device LLM.
// No network call, no data leaving the device.
//
// Covers (C# file -> Kotlin type):
//   TranslationTypes.cs      -> TranslationMode, TranslationRequest,
//                               TranslationResult, ConversationTurn
//   ITranslationEngine.cs    -> ITranslationEngine
//   ILiveTranslator.cs       -> ILiveTranslator
//   LlmTranslationEngine.cs  -> LlmTranslationEngine
//
// Fidelity notes:
//   * C# `record` -> Kotlin `data class`; `with { … }` -> `copy(...)`.
//   * C# `IAsyncEnumerable<T>` -> `kotlinx.coroutines.flow.Flow<T>`.
//   * C# `DateTimeOffset.UtcNow` -> `java.time.Instant.now()`.
//   * Backed by `inference.IChatGenerator` (already ported): `generateAsync`
//     returns the reply String; `streamAsync` returns `Flow<String>`.
//   * `ChatMessage` is `models.ChatMessage(id, role, content, createdAt)`; the
//     C# `new ChatMessage("user", text)` maps to a fresh id + role="user".
//   * `IsLanguagePairSupportedAsync` always returns `true` — the on-device LLM
//     handles any pair it was trained on.

package com.bhengubv.circleai.languagestranslation

import com.bhengubv.circleai.inference.IChatGenerator
import com.bhengubv.circleai.models.ChatMessage
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.time.Instant
import java.util.UUID

// =====================================================================
// TranslationTypes (TranslationTypes.cs)
// =====================================================================

enum class TranslationMode { Standard, Conversational, Document, Technical, Legal, Medical }

/** A request to translate a piece of text between two languages. */
data class TranslationRequest(
    val text: String,
    val sourceBcpTag: String,
    val targetBcpTag: String,
    val mode: TranslationMode = TranslationMode.Standard,
    val contextHint: String? = null,
)

/** Result of a completed translation. */
data class TranslationResult(
    val originalText: String,
    val translatedText: String,
    val sourceBcpTag: String,
    val targetBcpTag: String,
    val confidence: Float,
    val translatedAt: Instant,
)

/** One turn in a live bidirectional conversation. */
data class ConversationTurn(
    val speakerBcpTag: String,
    val originalText: String,
    val translatedText: String?,
    val timestamp: Instant,
)

// =====================================================================
// ITranslationEngine (ITranslationEngine.cs)
// =====================================================================

/**
 * On-device translation engine. No network call, no data leaving the device.
 * Translates meaning — not just words — using the on-device LLM.
 */
interface ITranslationEngine {
    suspend fun translate(request: TranslationRequest): TranslationResult

    fun streamTranslate(request: TranslationRequest): Flow<String>

    suspend fun isLanguagePairSupported(sourceBcpTag: String, targetBcpTag: String): Boolean
}

// =====================================================================
// ILiveTranslator (ILiveTranslator.cs)
// =====================================================================

/**
 * Bidirectional live conversation translator. Party A speaks [partyABcpTag];
 * party B speaks [partyBBcpTag]. Each turn is translated in real-time so both
 * parties hear each other. Runs entirely on-device. No API call. No data leaves
 * the device. Example: you speak Zulu, they hear English — and vice versa.
 */
interface ILiveTranslator : ITranslationEngine {
    fun streamConversation(
        inputStream: Flow<ConversationTurn>,
        partyABcpTag: String,
        partyBBcpTag: String,
    ): Flow<ConversationTurn>
}

// =====================================================================
// LlmTranslationEngine (LlmTranslationEngine.cs)
// =====================================================================

/**
 * [ITranslationEngine] backed by the on-device LLM via [IChatGenerator]. All
 * processing is on-device — no API calls, no data leaving the device.
 */
class LlmTranslationEngine(private val generator: IChatGenerator) : ILiveTranslator {

    override suspend fun translate(request: TranslationRequest): TranslationResult {
        val messages = listOf(userMessage(buildPrompt(request)))
        val translated = generator.generateAsync(messages)

        return TranslationResult(
            originalText = request.text,
            translatedText = translated.trim(),
            sourceBcpTag = request.sourceBcpTag,
            targetBcpTag = request.targetBcpTag,
            confidence = 0.9f,
            translatedAt = Instant.now(),
        )
    }

    override fun streamTranslate(request: TranslationRequest): Flow<String> = flow {
        val messages = listOf(userMessage(buildPrompt(request)))
        generator.streamAsync(messages).collect { token -> emit(token) }
    }

    override suspend fun isLanguagePairSupported(
        sourceBcpTag: String,
        targetBcpTag: String,
    ): Boolean = true // On-device LLM handles any pair it was trained on.

    override fun streamConversation(
        inputStream: Flow<ConversationTurn>,
        partyABcpTag: String,
        partyBBcpTag: String,
    ): Flow<ConversationTurn> = flow {
        inputStream.collect { turn ->
            val targetTag = if (turn.speakerBcpTag == partyABcpTag) partyBBcpTag else partyABcpTag

            val req = TranslationRequest(
                text = turn.originalText,
                sourceBcpTag = turn.speakerBcpTag,
                targetBcpTag = targetTag,
                mode = TranslationMode.Conversational,
            )

            val result = translate(req)
            emit(turn.copy(translatedText = result.translatedText))
        }
    }

    private companion object {
        fun userMessage(content: String): ChatMessage =
            ChatMessage(id = UUID.randomUUID().toString(), role = "user", content = content)

        fun buildPrompt(r: TranslationRequest): String =
            "Translate the following text from ${r.sourceBcpTag} to ${r.targetBcpTag}. " +
                "Mode: ${r.mode}. Preserve meaning and cultural context, not just literal words. " +
                (if (r.contextHint != null) "Context: ${r.contextHint}. " else "") +
                "Return only the translation with no explanation.\n\n${r.text}"
    }
}
